sys/can/dll.c

```diff
@@ -0,0 +1,274 @@
+
+    if (pkt == NULL) {
+        return -ENOMEM;
+    }
+
+    msg.type = CAN_MSG_SEND_FRAME;
+    msg.content.ptr = (void*) pkt;
+
+    msg_send_receive(&msg, &reply, candev_list[ifnum]);
+    if ((int)reply.content.value < 0) {
+        can_pkt_free(pkt);
+        return -ENOMEM;
+    }
+
+    return pkt->handle;
+}
+
+int raw_can_subscribe_rx(unsigned int ifnum, struct can_filter *filter, kernel_pid_t pid)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
+
+    msg_t msg, reply;
+
+    if (can_router_register(ifnum, pid, filter->can_id) < 0) {
+        return -ENOMEM;
+    }
+
+    msg.type = CAN_MSG_SET_FILTER;
+    msg.content.ptr = filter;
+    msg_send_receive(&msg, &reply, candev_list[ifnum]);
+
+    if ((int) reply.content.value < 0) {
+        can_router_unregister(ifnum, pid, filter->can_id);
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+
+int raw_can_unsubscribe_rx(unsigned int ifnum, struct can_filter *filter, kernel_pid_t pid)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (filter == NULL) {
+        return -ENOENT;
+    }
+
+    msg_t msg, reply;
+
+    if (can_router_unregister(ifnum, pid, filter->can_id) < 0) {
+        return -ENOMEM;
+    }
+
+    msg.type = CAN_MSG_REMOVE_FILTER;
+    msg.content.ptr = filter;
+    msg_send_receive(&msg, &reply, candev_list[ifnum]);
+
+    if ((int) reply.content.value < 0) {
+        return -ENOMEM;
+    }
+
+    return 0;
+}
+
+int raw_can_unsubscribe_rx_all(unsigned int ifnum, kernel_pid_t pid)
+{
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    msg_t msg, reply;
+    struct can_filter filter;
+
+    filter.can_mask = 0xffffffff;
+    msg.type = CAN_MSG_REMOVE_FILTER;
+    msg.content.ptr = &filter;
+
+    while (can_router_pop(ifnum, pid, &filter.can_id) == 0) {
+        msg_send_receive(&msg, &reply, candev_list[ifnum]);
+        if ((int) reply.content.value < 0) {
+            return -ENOMEM;
+        }
+    }
+
+    return 0;
+}
+
+int raw_can_free_frame(struct can_frame *frame)
+{
+    return can_router_free_frame(frame);
+}
+
+int raw_can_get_can_opt(unsigned int ifnum, can_opt_t *opt)
+{
+    msg_t msg, reply;
+
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (opt == NULL) {
+        return -ENOMEM;
+    }
+
+    opt->context = (uint16_t)candev_list[ifnum];
+
+    msg.type = CAN_MSG_GET;
+    msg.content.ptr = (void *)opt;
+    msg_send_receive(&msg, &reply, opt->context);
+
+    return (int) reply.content.value;
+}
+
+int raw_can_set_can_opt(unsigned int ifnum, can_opt_t *opt)
+{
+    msg_t msg, reply;
+
+    if (ifnum >= CAN_DLL_NUMOF) {
+        return -ENODEV;
+    }
+
+    if (opt == NULL) {
+        return -ENOMEM;
+    }
+
+    opt->context = (uint16_t)candev_list[ifnum];
+
+    msg.type = CAN_MSG_SET;
+    msg.content.ptr = (void *)opt;
+    msg_send_receive(&msg, &reply, opt->context);
```

Comment:
This is a copy of the architecture of gnrc_netdev2. And here the dll keep track of ifnum and pid, so it doesn't have a pointer to the device, then every operation from upper layer goes through the `_can_device_thread`.